iodata/formats/xyz.py

```diff
@@ -41,55 +71,99 @@
 PATTERNS = ['*.xyz']
 
 
-@document_load_one("XYZ", ['atcoords', 'atnums', 'title'])
-def load_one(lit: LineIterator) -> dict:
+DEFAULT_ATOM_COLUMNS = [
+    ("atnums", None, (), int,
+     (lambda word: int(word) if word.isdigit() else sym2num[word.title()]),
+     (lambda atnum: "{:2s}".format(num2sym[atnum]))),
+    ("atcoords", None, (3,), float,
+     (lambda word: float(word) * angstrom),
+     (lambda value: "{:15.10f}".format(value / angstrom)))
+]
+
+
+ATOM_COLUMNS_DOC = """\
+A list of atomic fields to be loaded. Each field as a tuple with the following
+items: **attribute** (``str``), **key** (``None`` or ``str``, when ``str`` the
+``IOData`` attribute is a ``dict``), **shape** for one atom (``tuple``),
+**dtype**, **load_word** (function taking string and returning a value with the
+correct type), **dump_word** (function taking a value and returning a formatted
+string).
+"""
+
+
+@document_load_one("XYZ", ['atcoords', 'atnums', 'title'],
+                   [], {"atom_columns": ATOM_COLUMNS_DOC})
+def load_one(lit: LineIterator, atom_columns=None) -> dict:
     """Do not edit this docstring. It will be overwritten."""
-    size = int(next(lit))
-    title = next(lit).strip()
-    atcoords = np.empty((size, 3), float)
-    atnums = np.empty(size, int)
-    for i in range(size):
+    if atom_columns is None:
+        atom_columns = DEFAULT_ATOM_COLUMNS
+    # Load the header
+    natom = int(next(lit))
+    data = {"title": next(lit).strip()}
+    # Load the atom lines.
+    for iatom in range(natom):
         words = next(lit).split()
-        try:
-            atnums[i] = sym2num[words[0].title()]
-        except KeyError:
-            atnums[i] = int(words[0])
-        atcoords[i, 0] = float(words[1]) * angstrom
-        atcoords[i, 1] = float(words[2]) * angstrom
-        atcoords[i, 2] = float(words[3]) * angstrom
-    return {
-        'title': title,
-        'atcoords': atcoords,
-        'atnums': atnums
-    }
-
-
-@document_load_many("XYZ", ['atcoords', 'atnums', 'title'])
-def load_many(lit: LineIterator) -> Iterator[dict]:
+        for attrname, keyname, shapesuffix, dtype, loadword, _dumpword in atom_columns:
+            if iatom == 0:
+                # In case of the first atom, the array with atomic properties
+                # is created and put in the right place.
+                dest = np.zeros((natom,) + shapesuffix, dtype=dtype)
+                if keyname is None:
+                    data[attrname] = dest
+                else:
+                    data.setdefault(attrname, {})[keyname] = dest
+                dest = dest[:1]
+            else:
+                # For all remaining atoms, the array with atomic properties is
+                # just retrieved.
+                if keyname is None:
```

Comment:
I´m not an expert in python, but it is hard to understand how this works. Could you explain it a little bit better?

Reply:
I've made a few changes and added more comments. Can you check if there is anything specific that is not obvious?